### PR TITLE
Simplify Networking to ProtoMulticastListener

### LINKED ISSUE #1697: Simplify Networking to Only ProtoMulticastListener
### Description of the task
With `primitive_google_to_nanopb_converter.h`, we can directly translate `google proto primitives` to `nanopb primitives`, which means that we can avoid separate `proto_multicast_listener`s and` nanopb_primitive_set_multicast_listener`s. With this ticket, we should collapse the hierarchy of listeners in `software/networking` and only use `proto_multicast_listener`s 

### Acceptance criteria

- [ ] Move common code from `multicast_listener `to `proto_multicast_listener`
- [ ] Delete `multicast_listener`, `nanopb_primitive_set_multicast_listener`
- [ ] Replace usages of `nanopb_primitive_set_multicast_listener` with `proto_multicast_listener` and a conversion

### PR DESCRIPTION
**Pull request by kelvinkoon**
### Description

Merge `multicast_listener` into `proto_multicast_listener`.  Replaced usages of nanopb_primitive_set_multicast_listener with proto_multicast_listener and a conversion to `nano_pb`.

### Testing Done

CI passes.

### Resolved Issues

Resolves #1697 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/simulation/standalone_simulator.cpp [hunk 1]
```diff
@@ -130,15 +131,17 @@ SSLProto::SSL_WrapperPacket StandaloneSimulator::getSSLWrapperPacket() const
 }
 
 void StandaloneSimulator::setYellowRobotPrimitives(
-    const TbotsProto_PrimitiveSet& primitive_set_msg)
+    const TbotsProto::PrimitiveSet& primitive_set_msg)
 {
-    simulator.setYellowRobotPrimitiveSet(primitive_set_msg);
+    auto nano_pb_primitive_set_msg = createNanoPbPrimitiveSet(primitive_set_msg);
+    simulator.setYellowRobotPrimitiveSet(nano_pb_primitive_set_msg);
```
Comment (jonathanlew) [inline src/software/simulation/standalone_simulator.cpp]:
```suggestion
    simulator.setYellowRobotPrimitiveSet(createNanoPbPrimitiveSet(primitive_set_msg));
```

File: src/software/simulation/standalone_simulator.cpp [hunk 2]
```diff
@@ -130,15 +131,17 @@ SSLProto::SSL_WrapperPacket StandaloneSimulator::getSSLWrapperPacket() const
...
 }
 
 void StandaloneSimulator::setBlueRobotPrimitives(
-    const TbotsProto_PrimitiveSet& primitive_set_msg)
+    const TbotsProto::PrimitiveSet& primitive_set_msg)
 {
-    simulator.setBlueRobotPrimitiveSet(primitive_set_msg);
+    auto nano_pb_primitive_set_msg = createNanoPbPrimitiveSet(primitive_set_msg);
+    simulator.setBlueRobotPrimitiveSet(nano_pb_primitive_set_msg);
```
Comment (jonathanlew) [inline src/software/simulation/standalone_simulator.cpp]:
```suggestion
    simulator.setBlueRobotPrimitiveSet(createNanoPbPrimitiveSet(primitive_set_msg));
```

File: src/software/networking/proto_multicast_listener.h [hunk 3]
```diff
@@ -28,17 +26,25 @@ class ProtoMulticastListener
...
+
+    // A UDP socket that we listen on for ReceiveProtoT messages from the network
+    boost::asio::ip::udp::socket socket_;
+    // The endpoint for the sender
+    boost::asio::ip::udp::endpoint sender_endpoint_;
 
-    // The multicast listener that receives the data and passes it to us via
-    // `handleDataReception`
-    MulticastListener multicast_listener;
+    static constexpr unsigned int max_buffer_length = 9000;
```
Comment (jonathanlew) [inline src/software/networking/proto_multicast_listener.h]: max_buffer_length should be ALL_CAPS

**Review by jonathanlew — CHANGES_REQUESTED** [review]
looks good, found a few nits

File: src/software/networking/threaded_proto_multicast_listener.tpp [hunk 4]
```diff
@@ -1,19 +1,19 @@
...
+    : io_service(), multicast_listener(io_service, ip_address, port, receive_callback)
 {
     // start the thread to run the io_service in the background
     io_service_thread = std::thread([this]() { io_service.run(); });
 }
 
-template <class ListenerT, typename CallBackFunctionArgT>
-ThreadedMulticastListener<ListenerT, CallBackFunctionArgT>::~ThreadedMulticastListener()
+template <class ReceiveProtoT>
+
```
Comment (akhilveeraghanta) [inline src/software/networking/threaded_proto_multicast_listener.tpp]: uber mega nit: extra newline here

File: src/software/networking/proto_multicast_listener.tpp [hunk 5]
```diff
@@ -8,26 +8,83 @@ template <class ReceiveProtoT>
...
+    }
+
+    if (num_bytes_received <= max_buffer_length)
+    {
+        LOG(WARNING)
+            << "num_bytes_received <= max_buffer_length, "
+            << "which means that the receive buffer is full and data loss has potentially occurred. "
+            << "Consider increasing max_buffer_length";
+    }
+}
```
Comment (akhilveeraghanta) [inline src/software/networking/proto_multicast_listener.tpp]: The logic here is inverted, we would want to log a warning when the num_bytes_received is greater than the max_buffer_length

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
Nice yeetage

File: src/software/simulation/standalone_simulator.cpp (hunk 2 above)
Comment (kelvinkoon) [reply src/software/simulation/standalone_simulator.cpp]: Done.

File: src/software/simulation/standalone_simulator.cpp (hunk 1 above)
Comment (kelvinkoon) [reply src/software/simulation/standalone_simulator.cpp]: Done.

File: src/software/networking/proto_multicast_listener.h (hunk 3 above)
Comment (kelvinkoon) [reply src/software/networking/proto_multicast_listener.h]: Done.

File: src/software/networking/proto_multicast_listener.tpp (hunk 5 above)
Comment (kelvinkoon) [reply src/software/networking/proto_multicast_listener.tpp]: Done.

File: src/software/networking/threaded_proto_multicast_listener.tpp (hunk 4 above)
Comment (kelvinkoon) [reply src/software/networking/threaded_proto_multicast_listener.tpp]: uber mega done.

**Review by akhilveeraghanta — APPROVED** [review]
LGTM! Nice cleanup of our networking libraries

**Review by jonathanlew — APPROVED** [review]

**Pull request merged**